[RFC] prov/verbs: Allow selective send completions

patrickmacarthur opened the pull request:
Note that I have not had a chance to test this PR, since my testing of this also relies on FI_INJECT support which the verbs provider also lacks (and which is somewhat more complicated to add).

This commit adds support for the FI_COMPLETION flag for binding a
send completion queue to an endpoint and for selectively enabling
completions.

See issue #775.

Signed-off-by: Patrick MacArthur <pmacarth@iol.unh.edu>

shefty commented:
@a-ilango - can you let me know if this looks okay?

@patrickmacarthur - it looks like this PR needs to be refreshed.

patrickmacarthur commented:
I have rebased the PR onto master.

a-ilango commented on prov/verbs/src/fi_verbs.c:
```diff
@@ -872,6 +881,8 @@ fi_ibv_msg_ep_sendmsg(struct fid_ep *ep, const struct fi_msg *msg, uint64_t flag
 		wr.send_flags = 0;
 	}
+	if ((_ep->ep_flags & flags & FI_COMPLETION))
+		wr.send_flags |= IBV_SEND_SIGNALED;
```
If ep_flags is not set then we have to generate completion by default. Same comment applies for writemsg, readmsg and atomics below.

patrickmacarthur commented:
Ok, I will fix these issues and re-submit.

patrickmacarthur commented:
I have pushed an updated version which addresses the issues that you raised.  In addition to implementing support for `FI_COMPLETION` in `tx_attr->op_flags`, I also added `FI_INJECT` and `FI_REMOTE_COMPLETE` while I was at it.  The former because it was straightforward, the latter because it is required by the `ft_finalize()` function in fabtests.

shefty commented:
@a-ilango - are we good to merge this?

a-ilango commented:
>This commit also eliminates the behavior where the provider
>automatically sends small messages inline, since this feature can now be
>controlled entirely by the user.

Sending small messages inline would help decrease latency. So may be we should retain the behavior     even though the user has the control. @shefty Comments?

a-ilango commented on prov/verbs/src/fi_verbs.c:
```diff
@@ -1132,12 +1186,18 @@ fi_ibv_msg_ep_rma_readmsg(struct fid_ep *ep, const struct fi_msg_rma *msg,
 
 	_ep = container_of(ep, struct fi_ibv_msg_ep, ep_fid);
+	if (!(_ep->ep_flags & FI_COMPLETION)
+			|| (flags & (FI_COMPLETION|FI_REMOTE_COMPLETE))) {
```
FI_REMOTE_COMPLETE does not apply to readmsg.

shefty commented:
If the provider can give better performance, I vote for it.  So, I would allow the use of inline, even if the app does not request it.  The drawback is that it does introduce a branch into the code.

patrickmacarthur commented:
I have a revised branch that addresses both issues that were raised with this PR  which I will push once I have tested it.
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

patrickmacarthur commented:
Tested and pushed an updated changeset which addresses the issues raised by @a-ilango.  This was tested with a heavily modified `simple/fi_msg_rma` from fabtests which requested completion for RMA write requests only periodically, as well as using an EXS test program.

a-ilango commented:
The updated changes look fine. Thanks @patrickmacarthur !

I saw a new patch that changes the return value in case of QP overflows. Was that intended to be part of this PR?

patrickmacarthur commented:
That was something that I discovered while testing this patchset.  I can separate it into a different PR if you would like.

shefty commented on prov/verbs/src/fi_verbs.c:
```diff
@@ -844,6 +875,15 @@ fi_ibv_msg_ep_recvv(struct fid_ep *ep, const struct iovec *iov, void **desc,
+{
+	int ret;
+	ret = ibv_post_send(qp, wr, bad_wr);
+	return (ret == ENOMEM) ? -FI_EAGAIN : -ret;
```
The checks for ibv_post_send/recv should be against -1, with errno checked for ENOMEM. I.e.
if (ret)
    return (errno == ENOMEM) ? -FI_EAGAIN : -errno;
else
   return 0;

patrickmacarthur replied:
I've removed this from the PR and will submit a separate PR for this, so that we can discuss it separately.

patrickmacarthur replied:
Submitted this commit separately as PR #824.

shefty commented on prov/verbs/src/fi_verbs.c:
```diff
@@ -78,6 +78,7 @@
 		FI_SEND | FI_RECV | FI_REMOTE_READ | FI_REMOTE_WRITE | \
 		FI_REMOTE_SIGNAL)
 #define VERBS_MODE (FI_LOCAL_MR | FI_PROV_MR_ATTR)
+#define VERBS_TX_OP_FLAGS (FI_INJECT | FI_COMPLETION | FI_REMOTE_COMPLETE)
```
I already merged this PR, but this is wrong.  There should not be any default op_flags defined by the provider.  These should only be set by the app.  FI_INJECT is doubly wrong, in that it forces the use of inject for all operations.

shefty replied:
ignore this comment - this is simply listing the supported values for checking later
